x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamIndexTransportAction.java

```diff
@@ -0,0 +1,240 @@
+
+    private static final Set<String> SETTINGS_TO_ADD_BACK = Set.of(IndexMetadata.SETTING_BLOCKS_WRITE, IndexMetadata.SETTING_READ_ONLY);
+
+    private static final IndicesOptions IGNORE_MISSING_OPTIONS = IndicesOptions.fromOptions(true, true, false, false);
+    private final ClusterService clusterService;
+    private final Client client;
+    private final IndexScopedSettings indexScopedSettings;
+
+    @Inject
+    public ReindexDataStreamIndexTransportAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ActionFilters actionFilters,
+        Client client,
+        IndexScopedSettings indexScopedSettings
+    ) {
+        super(
+            ReindexDataStreamIndexAction.NAME,
+            false,
+            transportService,
+            actionFilters,
+            ReindexDataStreamIndexAction.Request::new,
+            transportService.getThreadPool().executor(ThreadPool.Names.GENERIC)
+        );
+        this.clusterService = clusterService;
+        this.client = client;
+        this.indexScopedSettings = indexScopedSettings;
+    }
+
+    @Override
+    protected void doExecute(
+        Task task,
+        ReindexDataStreamIndexAction.Request request,
+        ActionListener<ReindexDataStreamIndexAction.Response> listener
+    ) {
+        var sourceIndexName = request.getSourceIndex();
+        var destIndexName = generateDestIndexName(sourceIndexName);
+        IndexMetadata sourceIndex = clusterService.state().getMetadata().index(sourceIndexName);
+        Settings settingsBefore = sourceIndex.getSettings();
+
+        if (sourceIndex.getCreationVersion().isLegacyIndexVersion() == false) {
+            logger.warn(
+                "Reindexing datastream index with non-legacy version, index [{}], version [{}]",
+                sourceIndexName,
+                sourceIndex.getCreationVersion()
+            );
+        }
+
+        AtomicReference<BulkByScrollResponse> reindexResponse = new AtomicReference<>();
+        SubscribableListener.<AcknowledgedResponse>newForked(l -> setBlockWrites(sourceIndexName, l))
+            .<AcknowledgedResponse>andThen(l -> deleteDestIfExists(request.getDeleteDestIfExists(), destIndexName, l))
+            .<CreateIndexResponse>andThen(l -> createIndex(sourceIndex, destIndexName, l))
+            .<BulkByScrollResponse>andThen(l -> reindex(sourceIndexName, destIndexName, l))
+            .andThenApply(r -> {
+                reindexResponse.set(r);
+                return r;
+            })
+            .<AcknowledgedResponse>andThen(l -> updateSettings(settingsBefore, destIndexName, l))
+            .andThenApply(ignored -> new ReindexDataStreamIndexAction.Response(destIndexName, reindexResponse.get().getCreated()))
+            .addListener(listener);
+    }
+
+    private void setBlockWrites(String sourceIndexName, ActionListener<AcknowledgedResponse> listener) {
```

Comment:
Should we remove the write block on the source index after reindexing? I'm inclined to say it's fine to leave it, but if we failed/cancelled an instance of this call then reran it, the block-writes setting would then be found on the source index and copied to the new dest index.

Reply:
We discussed adding a flag to the cluster state which says whether an index had block-write before the call to this action. This would then be passed as a param into the action.